tinyfilemanager.php

```diff
@@ -1053,6 +1068,75 @@ function get_file_path () {
     exit();
 }
 
+// Mass downloading
+if (isset($_POST['group'], $_POST['download'], $_POST['token']) && !FM_READONLY) {
+
+    // Verify token to ensure it's valid
+    if (!verifyToken($_POST['token'])) {
+        fm_set_msg(lng("Invalid Token."), 'error');
+        exit;
+    }
+
+    $path = FM_ROOT_PATH;
+    if (FM_PATH != '') {
+        $path .= '/' . FM_PATH;
+    }
+
+    $errors = 0;
+    $files = $_POST['file']; // List of selected files
+    if (is_array($files) && count($files)) {
+
+        // Create a new ZIP archive
+        $zip = new ZipArchive();
+        $zip_filename = 'download_' . date('Y-m-d_H-i-s') . '.zip';
+        $zip_filepath = sys_get_temp_dir() . '/' . $zip_filename;
+
+        if ($zip->open($zip_filepath, ZipArchive::CREATE) !== TRUE) {
+            fm_set_msg(lng('Cannot create ZIP file'), 'error');
+            $FM_PATH=FM_PATH; fm_redirect(FM_SELF_URL . '?p=' . urlencode($FM_PATH));
+        }
+
+        foreach ($files as $f) {
+            if ($f != '') {
+                $new_path = $path . '/' . fm_clean_path($f); // Sanitize the file path
+
+                if (is_file($new_path)) {
+                    // Add the file to the ZIP archive
+                    $zip->addFile($new_path, basename($new_path));
+                } else {
+                    $errors++;
```

Comment:
It clearly does not work with folders, so prompting `Download selected files and folders?` is misleading.

Reply:
@vadasziattila can you add folder as well.

Reply:
I think here is a good opportunity to make good use of the included `FM_Zipper` class.
Just a suggestion, but this would work okay:
```php
// Mass downloading
if (isset($_POST['group'], $_POST['download'], $_POST['token']) && !FM_READONLY) {

    // Verify token to ensure it's valid
    if(!verifyToken($_POST['token'])) {
        fm_set_msg(lng("Invalid Token."), 'error');
        exit;
    }

    $path = FM_ROOT_PATH;
    if (FM_PATH != '') {
        $path .= '/' . FM_PATH;
    }

    $files = $_POST['file']; // List of selected files
    if (is_array($files) && count($files)) {
        // Create a new ZIP archive
        $zip = new FM_Zipper();
        $zip_filename = 'download_' . date('Y-m-d_H-i-s') . '.zip';
        $zip_filepath = sys_get_temp_dir() . '/' . $zip_filename;

        chdir($path);
        $sanitized_filepaths = array();
        foreach($files as $f){
            if ($f != '') {
                array_push($sanitized_filepaths, fm_clean_path($f)); // Sanitize the file path
            }
        }
        $res = $zip->create($zip_filepath, $sanitized_filepaths);

        if ($res && file_exists($zip_filepath)) {
            // Serve the ZIP file for download
            header('Content-Type: application/zip');
            header('Content-Disposition: attachment; filename="' . $zip_filename . '"');
            header('Content-Length: ' . filesize($zip_filepath));
            readfile($zip_filepath);
            // Remove the ZIP file from the temporary directory after download
            unlink($zip_filepath);
            exit;
        } else {
            fm_set_msg(lng('Error creating ZIP file'), 'error');
        }
    } else {
        fm_set_msg(lng('Nothing selected'), 'alert');
    }

    $FM_PATH = FM_PATH;
    fm_redirect(FM_SELF_URL . '?p=' . urlencode($FM_PATH));
}
```